File: frontend/src/api/agents.ts
import { apiGet, apiPost, apiPut } from "./client";

export interface AgentTools {
  generate_image: boolean;
  generate_video: boolean;
  image_to_image: boolean;
  analyze_image: boolean;
  search_inspiration: boolean;
}

// 多 Agent 预设：一套人设/行为配置，对话时可切换
export interface Agent {
  id: string;
  name: string;
  systemPrompt: string;
  memory: string;
  temperature: number | null;
  topP: number | null;
  maxTokens: number | null;
  tools: AgentTools;
  mcpServerIds: string[];   // 选中启用的 MCP 服务器 id（空=都不用）
  skillIds: string[];       // 选中启用的技能 id（空=都不用）
  isDefault: boolean;
  enabled: boolean;
}

export function listAgents() {
  return apiGet<Agent[]>("/agents");
}

export function saveAgents(agents: Agent[]) {
  return apiPost<Agent[]>("/agents", agents);
}

// 内置默认系统提示词（普通对话优先 + 显式工具调用规则）
export function defaultPrompt() {
  return apiGet<{ prompt: string }>("/agents/default-prompt");
}

// ③ 内置智能体：图里所有默认 Agent 的元数据 + 默认值 + 当前生效值（含覆盖）
export type BuiltinKind = "llm" | "rules" | "specialist";
export interface BuiltinAgent {
  id: string;
  name: string;
  kind: BuiltinKind;
  role: string;
  tools: string[];
  editable: string[];              // 可覆盖字段名（llm: systemPrompt/temperature；rules: gateFloor/gateBaseRate/tiers）
  defaults: Record<string, unknown>;
  effective: Record<string, unknown>;  // 默认叠加用户覆盖后的生效值
}
// 覆盖表：{agent_id: {field: value}}
export type BuiltinOverrides = Record<string, Record<string, unknown>>;

export function listBuiltinAgents() {
  return apiGet<BuiltinAgent[]>("/agents/builtin");
}

export function saveBuiltinOverrides(overrides: BuiltinOverrides) {
  return apiPost<BuiltinAgent[]>("/agents/builtin", overrides);
}

// ④ 固化知识库（DATA_DIR/agent_knowledge/*.md，智能编造 Agent 规范）
export interface KnowledgeDoc {
  name: string;      // 文件主名（展示/读取用）
  file: string;      // 完整文件名（xxx.md）
  size: number;      // 字节
  mtime: number;     // 毫秒时间戳
  truncated: boolean; // 超出单文档注入/预览上限
  // 以下仅带 frontmatter（skill）的技能文档才有；普通知识文档无这些键
  skill?: string;    // 技能名（kebab-case，smart 模式按此触发装载）
  whenToUse?: string; // 触发描述（注入目录时展示，供模型判断何时 load_doc）
  tools?: string[];  // 工具清单（skill 可调用的能力）
}
export interface KnowledgeDocContent extends KnowledgeDoc {
  content: string;   // 正文（技能文档=去 frontmatter；普通知识=全文；超长截断到单文档上限）
}

// 注入模式配置：smart=技能目录+按需 load_doc（默认）；always=全量常驻（老行为）
export type KnowledgeMode = "smart" | "always";
export interface KnowledgeConfig {
  mode: KnowledgeMode;
  always_docs: string[];   // smart 模式下强制常驻注入的技能文档主名
}

export function listAgentKnowledge() {
  return apiGet<KnowledgeDoc[]>("/agents/knowledge");
}

export function readAgentKnowledge(name: string) {
  return apiGet<KnowledgeDocContent>(`/agents/knowledge/${encodeURIComponent(name)}`);
}

export function getKnowledgeConfig() {
  return apiGet<Partial<KnowledgeConfig>>("/agents/knowledge/config");
}

export function saveKnowledgeConfig(cfg: KnowledgeConfig) {
  return apiPut<KnowledgeConfig>("/agents/knowledge/config", cfg);
}

export const DEFAULT_TOOLS: AgentTools = {
  generate_image: true, generate_video: true, image_to_image: true, analyze_image: true,
  search_inspiration: true,
};

File: backend/app/routers/agents.py
"""多 Agent 预设端点：列表/保存 + 取内置默认提示词（供前端"默认规则"展示）。"""
from __future__ import annotations

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from app.services import agent_knowledge, agent_store, builtin_agents

router = APIRouter()


class AgentTools(BaseModel):
    generate_image: bool = True
    generate_video: bool = True
    image_to_image: bool = True
    analyze_image: bool = True
    search_inspiration: bool = True


class Agent(BaseModel):
    id: str = ""
    name: str
    systemPrompt: str = ""
    memory: str = ""
    temperature: float | None = None
    topP: float | None = None
    maxTokens: int | None = None
    tools: AgentTools = AgentTools()
    mcpServerIds: list[str] = []   # 选中启用的 MCP 服务器 id（空=都不用）
    skillIds: list[str] = []       # 选中启用的技能 id（空=都不用）
    isDefault: bool = False
    enabled: bool = True


@router.get("")
def list_agents() -> list[Agent]:
    return [Agent(**a) for a in agent_store.load_agents()]


@router.post("")
def save_agents(agents: list[Agent]) -> list[Agent]:
    saved = agent_store.save_agents([a.model_dump() for a in agents])
    return [Agent(**a) for a in saved]


@router.get("/default-prompt")
def default_prompt() -> dict:
    """返回内置默认系统提示词（普通对话优先 + 显式工具调用规则）。"""
    from app.services.image_agent import _AGENT_SYSTEM_BASE
    return {"prompt": _AGENT_SYSTEM_BASE}


# ── ③ 内置智能体：展示图里所有默认 Agent + 参数 + 绑定工具，开放高价值字段覆盖 ──


@router.get("/builtin")
def list_builtin_agents() -> list[dict]:
    """列出所有内置 Agent 的元数据 + 默认值 + 当前生效值（含用户覆盖）。"""
    return builtin_agents.registry_view()


@router.post("/builtin")
def save_builtin_overrides(overrides: dict) -> list[dict]:
    """保存内置 Agent 覆盖（{agent_id: {field: value}}），只接受注册表声明的可编辑字段。"""
    builtin_agents.save_overrides(overrides)
    return builtin_agents.registry_view()


# ── ④ 固化知识库：设置页「固化知识库（智能编造流程规范）」只读展示 + 注入模式配置 ──


@router.get("/knowledge")
def list_knowledge_docs() -> list[dict]:
    """列出 agent_knowledge/ 下的固化知识文档元数据（文件名升序）。"""
    return agent_knowledge.list_docs()


class KnowledgeConfig(BaseModel):
    mode: str = "smart"            # smart=技能按需装载（默认）/ always=全量常驻注入
    always_docs: list[str] = []    # smart 模式下强制常驻注入的技能文档主名


@router.get("/knowledge/config")
def read_knowledge_config() -> dict:
    """读取固化知识注入模式配置（settings.agentKnowledge，缺省 smart）。"""
    return agent_knowledge.injection_config()


@router.put("/knowledge/config")
def save_knowledge_config(cfg: KnowledgeConfig) -> dict:
    """保存固化知识注入模式配置（读-改-写 user_state，保留其它 settings 键）。"""
    if cfg.mode not in (agent_knowledge.MODE_SMART, agent_knowledge.MODE_ALWAYS):
        raise HTTPException(status_code=400, detail="mode 仅支持 smart/always")
    return agent_knowledge.save_injection_config(
        {"mode": cfg.mode, "always_docs": cfg.always_docs})


@router.get("/knowledge/{name}")
def read_knowledge_doc(name: str) -> dict:
    """按文件主名读取固化知识文档正文（只读，*.md 白名单 + 防路径穿越）。"""
    try:
        return agent_knowledge.read_doc(name)
    except FileNotFoundError as exc:
        raise HTTPException(status_code=404, detail="知识文档不存在") from exc
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
